crates/astria-sequencer/src/app/vote_extension.rs

```diff
@@ -0,0 +1,649 @@
+    debug,
+    trace,
+};
+
+use crate::{
+    address::StateReadExt as _,
+    authority::StateReadExt as _,
+    slinky::oracle::{
+        currency_pair_strategy::DefaultCurrencyPairStrategy,
+        state_ext::StateWriteExt,
+    },
+    state_ext::StateReadExt,
+};
+
+// https://github.com/skip-mev/slinky/blob/793b2e874d6e720bd288e82e782502e41cf06f8c/abci/types/constants.go#L6
+const MAXIMUM_PRICE_BYTE_LEN: usize = 33;
+
+pub(crate) struct Handler {
+    // gRPC client for the slinky oracle sidecar.
+    oracle_client: Option<OracleClient<Channel>>,
+}
+
+impl Handler {
+    pub(crate) fn new(oracle_client: Option<OracleClient<Channel>>) -> Self {
+        Self {
+            oracle_client,
+        }
+    }
+
+    pub(crate) async fn extend_vote<S: StateReadExt>(
+        &mut self,
+        state: &S,
+    ) -> anyhow::Result<abci::response::ExtendVote> {
+        let Some(oracle_client) = self.oracle_client.as_mut() else {
+            // we allow validators to *not* use the oracle sidecar currently,
+            // so this will get converted to an empty vote extension when bubbled up.
+            //
+            // however, if >1/3 of validators are not using the oracle, the prices will not update.
+            bail!("oracle client not set")
+        };
+
+        // if we fail to get prices within the timeout duration, we will return an empty vote
+        // extension to ensure liveness.
+        let prices = match oracle_client.prices(QueryPricesRequest {}).await {
+            Ok(prices) => prices.into_inner(),
+            Err(e) => {
+                bail!("failed to get prices from oracle sidecar: {e}",);
+            }
+        };
+
+        tracing::debug!(
+            prices_count = prices.prices.len(),
+            "got prices from oracle sidecar; transforming prices"
+        );
+
+        let oracle_vote_extension = transform_oracle_service_prices(state, prices)
+            .await
+            .context("failed to transform oracle service prices")?;
+
+        Ok(abci::response::ExtendVote {
+            vote_extension: oracle_vote_extension.into_raw().encode_to_vec().into(),
+        })
+    }
+
+    pub(crate) async fn verify_vote_extension<S: StateReadExt>(
+        &self,
+        state: &S,
+        vote: abci::request::VerifyVoteExtension,
+    ) -> abci::response::VerifyVoteExtension {
+        if vote.vote_extension.is_empty() {
+            return abci::response::VerifyVoteExtension::Accept;
+        }
+
+        match verify_vote_extension(state, vote.vote_extension, false).await {
+            Ok(()) => abci::response::VerifyVoteExtension::Accept,
+            Err(e) => {
+                tracing::error!(error = %e, "failed to verify vote extension");
+                abci::response::VerifyVoteExtension::Reject
+            }
+        }
+    }
+}
+
+// see https://github.com/skip-mev/slinky/blob/5b07f91d6c0110e617efda3f298f147a31da0f25/abci/ve/utils.go#L24
+async fn verify_vote_extension<S: StateReadExt>(
```

Comment:
OK, looking at how this is used: this function performs conflates normal expected operations (failing to decode bytes that are sent from external entities, ensuring that passed in data is correct), with potentially(?) catastrophic issues: getting information form state that should be there but isn't.

Reply:
will separate out the state read